src/Akka.Streams.Kafka/Stages/Consumers/Abstract/SubSourceLogic.cs

```diff
@@ -0,0 +1,275 @@
+            _subsourceCancelledCallback = GetAsyncCallback<(TopicPartition, Option<ConsumeResult<K, V>>)>(HandleSubsourceCancelled);
+            _subsourceStartedCallback = GetAsyncCallback<(TopicPartition, TaskCompletionSource<Done>)>(HandleSubsourceStarted);
+            
+            SetHandler(shape.Outlet, onPull: EmitSubSourcesForPendingPartitions, onDownstreamFinish: PerformShutdown);
+        }
+
+        public override void PostStop()
+        {
+            ConsumerActor.Tell(new KafkaConsumerActorMetadata.Internal.Stop(), SourceActor.Ref);
+            
+            OnShutdown();
+            
+            base.PostStop();
+        }
+
+        protected override void OnTimer(object timerKey)
+        {
+            if (timerKey is CloseRevokedPartitions)
+            {
+                Log.Debug("#{} Closing SubSources for revoked partitions: {}", _actorNumber, _partitionsToRevoke.JoinToString(", "));
+
+                _onRevoke(_partitionsToRevoke);
+                _pendingPartitions = _pendingPartitions.Except(_partitionsToRevoke);
+                _partitionsInStartup = _partitionsInStartup.Except(_partitionsToRevoke);
+                _partitionsToRevoke.ForEach(tp => _subSources[tp].SetResult(Done.Instance));
+                _subSources = _subSources.RemoveRange(_partitionsToRevoke);
+                _partitionsToRevoke = _partitionsToRevoke.Clear();
+            }
+        }
+
+        private async void HandlePartitionsAssigned(IImmutableSet<TopicPartition> assigned)
+        {
+            var formerlyUnknown = assigned.Except(_partitionsToRevoke);
+
+            if (Log.IsDebugEnabled && formerlyUnknown.Any())
+            {
+                Log.Debug("#{} Assigning new partitions: {}", _actorNumber, formerlyUnknown.JoinToString(", "));
+            }
+            
+            // make sure re-assigned partitions don't get closed on CloseRevokedPartitions timer
+            _partitionsToRevoke = _partitionsToRevoke.Except(assigned);
+
+            if (!_getOffsetsOnAssign.HasValue)
+            {
+                UpdatePendingPartitionsAndEmitSubSources(formerlyUnknown);
+            }
+            else
+            {
+                try
+                {
+                    var offsets = await _getOffsetsOnAssign.Value(assigned);
+                    
+                    SeekAndEmitSubSources(formerlyUnknown, offsets);
+                }
+                catch (Exception ex)
+                {
+                    _stageFailCallback(new ConsumerFailed($"{_actorNumber} Failed to fetch offset for partitions: {formerlyUnknown.JoinToString(", ")}", ex));
+                }
+            }
+        }
+        
+        private async void SeekAndEmitSubSources(IImmutableSet<TopicPartition> formerlyUnknown, IImmutableSet<TopicPartitionOffset> offsets)
+        {
+            try
+            {
+                await ConsumerActor.Ask(new KafkaConsumerActorMetadata.Internal.Seek(offsets), TimeSpan.FromSeconds(10));
```

Comment:
@Aaronontheweb alpakka passes `SourceActor.Ref` as a third parameter in `Ask` call here:
```scala
consumerActor
      .ask(KafkaConsumerActor.Internal.Seek(offsets))(Timeout(10.seconds), sourceActor.ref)
      .map(_ => updatePendingPartitionsAndEmitSubSourcesCb.invoke(formerlyUnknown))
      .recover {
        case _: AskTimeoutException =>
          stageFailCB.invoke(
            new ConsumerFailed(
              s"#$actorNumber Consumer failed during seek for partitions: ${offsets.keys.mkString(", ")}."
            )
          )
      }
```

`Akka.NET` does not allow passing sender ref to `Ask` method, and actually I do not understand why this might be used at all. Internal actor will be created, wait for message and return answer as a task result - all should be fine, and code above should work as expected.

Anyway, consumer actor handles `Seek` message with just `_consumer.Seek` and `Sender.Tell(Done)` calls, so I think there is no point in setting sender reference anyway (do special response for handling is expected), but wanted to share this with you.

Reply:
> Akka.NET does not allow passing sender ref to Ask method, and actually I do not understand why this might be used at all. Internal actor will be created, wait for message and return answer as a task result - all should be fine, and code above should work as expected.

That's good to know. Might be a consequence of some of the changes they've made to the Akka.Streams materializer in the JVM. But as long as the tests pass here I think we'll be fine.